cosmpy/aerial/coins.py

```diff
@@ -20,58 +20,295 @@
+
+        result_inout.clear()
+        result_inout.extend(res_dict.values())
+        result_inout.canonicalise()
+
+
+def parse_coins(value: str) -> List[CoinProto]:
     """Parse the coins.
 
-    :param value: coins
-    :raises RuntimeError: If unable to parse the value
-    :return: coins
+    :param value: coins encoded in cosmos-sdk string format
+    :return: List of CoinProto objects
     """
-    coins = []
+    return Coins(value).to_proto()
 
-    parts = re.split(r",\s*", value)
-    for part in parts:
-        part = part.strip()
-        if part == "":
-            continue
 
-        match = re.match(r"^(\d+)(.+)$", part)
-        if match is None:
-            raise RuntimeError(f"Unable to parse value {part}")
+CoinsParamType = Union[str, Coins, List[Coin], List[CoinProto], Coin, CoinProto]
 
-        # extract out the groups
-        amount, denom = match.groups()
-        coins.append(Coin(amount=amount, denom=denom))
 
-    return coins
+def is_denom_valid(denom: str) -> bool:
+    """Return true if coin denom name is valid.
 
+    :param denom: string denom
+    :return: bool validity
+    """
+    return denom_regex.match(denom) is not None
 
-CoinsParamType = Union[str, Coin, List[Coin]]
 
+# def is_coins_sorted(coins: Union[Coins, CoinsProto, List[Coin], List[CoinProto]]) -> bool:
+def is_coins_sorted(coins: Union[str, Coins, List[Coin], List[CoinProto]]) -> bool:
+    """Return true if given coins representation is sorted.
 
-def to_coins(amount: CoinsParamType) -> List[Coin]:
+    :param coins: Any type representing coins
+    :return: bool is_sorted
     """
-    Convert various fee amount formats into a standardized list of Coin objects.
+    if not coins:
+        return True
+
+    if isinstance(coins, str):
+        coins = Coins(coins)
+
+    seen = set()
+
+    last_denom = coins[0].denom
+    seen.add(last_denom)
+
+    for c in coins[1:]:
+        if last_denom >= c.denom:
+            return False
+
+    return True
+
+
+# def validate_coins(coins: Union[str, Coins, CoinsProto, List[Coin], List[CoinProto]]):
+def validate_coins(coins: Union[str, Coins, List[Coin], List[CoinProto]]):
+    """Return true if given coins representation is valid.
+
+    :param coins: Any type representing coins
+    :raises ValueError: If there are multiple coins with same denom or coins are not sorted alphabetically
+    :return: bool validity
+    """
+    if not coins:
+        return
+
+    if isinstance(coins, str):
+        coins = Coins(coins)
+
+    if len(coins) == 0:
+        return
+
+    seen = set()
+
+    last_denom = coins[0].denom
+    seen.add(last_denom)
+
+    for c in coins[1:]:
+        if c.denom not in seen:
+            raise ValueError(f'Multiple occurrences of the "{c.denom}" denomination')
+
+        if last_denom >= c.denom:
+            raise ValueError(
+                "Coins are not sorted as cosmos-sdk expects it (ascending based on denom)"
+            )
```

Comment:
This also looks inconsistent. Either the inequality should be flipped or `ascending` in the error message should be `descending`. Probably the latter if the description in line 105 is correct.

Reply:
IT should be `ascending` like `a,b,c`, that's why there is `reverse=False`